### ISSUE TITLE
Query statement not appearing in span

### ISSUE THREAD

**Issue by jcmackie** [justindsmith/opentelemetry-instrumentations-js#22]
Hello @justindsmith,

Sorry if this is not the right place for this kind of message.

After https://github.com/justindsmith/opentelemetry-instrumentations-js/pull/21 was merged and I pulled the latest version, I was disheartened to see that query statements were not appearing the spans.

After a couple of hours debugging, I found out that this is because that particular argument `query` only shows up in Raw type Prisma queries. And that for the regular queries (eg `findMany`) the arguments passed to the request function are not very helpful...
I spent a bit of time trying to find somewhere to grab the information from, but I was quite out of my depth as a non-Node programmer.
I also noticed that the Prisma project has started incorporating OTel directly into their project: https://github.com/prisma/prisma/tree/main/packages/instrumentation
But it's only in a 'preview' stage and quite useless at the moment.

For example, here is a span from their instrumentation:
```
{
   traceId: '62d8fa8c6196d98ee05098e643000b85',
   parentId: '5440f7cef2b3f2db',
   name: 'request',
   id: '8cce709152487426',
   kind: 0,
   timestamp: 1658387085373138,
   duration: 11629,
   attributes: {},
   status: { code: 0 },
   events: [],
   links: []
}
```

But given the fact that they are building it themselves... it might be better to direct further development effort into their codebase?
I'm still going to use yours for now because as it currently stands it's 100 times more useful.

Regards,
James

**Comment by justindsmith** [comment]
Hey @jcmackie, first, yeah, this project started as a temporary instrumentation while Prisma worked on an "official" opentelemetry integration. I've been following this Prisma issue to track their progress: https://github.com/prisma/client-planning/issues/21. As you found out, it's pretty bare bones right now. Once they have better support the plan is to deprecate the `opentelemetry-instrumentation-prisma-client` in favor of the official one. Good reminder that I should add a note to the plugin readme about this so it's more clear the plans.

For the `query` of the non-raw calls, I tried the same thing you did early on but didn't find a way to easily get the sql query so I abandoned the effort. Is there a specific thing you're trying to use that query for? In my mind, the sql query generated by the prisma client is effectively a "black box" implementation detail. But I _could_ see value in knowing the effect of different parameters on the query to understand which calls are the most expensive. For example, knowing that you're requesting a limit of 10000 instead of 10 might be useful information to know why something is taking so long.

Instead of using the raw sql as the `SemanticAttributes.DB_STATEMENT`, what if we just `JSON.stringified` the parameters passed into the prisma model call and used _that_ output as the statement? Or maybe created some kind of union of the `clientMethod` and parameters. Something like this:

```
notes.findMany {
  where: {
    author: "jcmackie"
  },
  take: 1000
}
```

Would that be helpful?

Let me know if there is a specific use case where knowing the raw sql is the important though.

**Comment by jcmackie** [comment]
Hi @justindsmith 

Thanks for replying, and sorry for the delay.

> Is there a specific thing you're trying to use that query for? In my mind, the sql query generated by the prisma client is effectively a "black box" implementation detail.

That's a good question. I'm tracking down some performance 'issues' in an app that uses Prisma. There were two reasons I wanted to know the actual SQL query(ies) that are being executed:

1. It would allow me to manually run the query behind an `EXPLAIN` statement, which in turns provides some valuable insight into performance problems with the query or schema.
2. It would show subquery/nested query explosions. One of the situations I was tracking, was a case where because of relations between tables shown in the `Prisma.schema`, it was doing fairly large nested queries. To be honest I could have found this by studying the code and `Prisma.schema` more closely, but seeing the multiple queries being executed as separate spans would have made it more obvious. But I am honestly not sure if that is the right way to instrument an ORM, because as you say, it should all be an implementation detail. But I have found myself caught out several times by the weird things that ORMs do.

> Would that be helpful?

I think it might be helpful for sure. :)

**Comment by justindsmith** [comment]
Okay, probably more of an advanced use case, but understanding the Prisma internals to better understand the effects of your usage and areas / queries to optimize does make sense.

Since each call to the model _could_ translate to multiple queries being performed, we'll likely need to track these internal calls as separate spans (as you call out). Let me see if there's another function we could patch which might give us access to the raw query being executed. Since you've peeked more into the Prisma code, if you know where that query is formed / executed and can give me a pointer that would be helpful!

**Comment by justindsmith** [comment]
I took a quick look and I'm not sure we can get the raw query like you're looking for using the patching approach this library (`opentelemetry-plugin-prisma-client`) is using.

The sql generation / execution is happening inside the [prisma/prisma-engines](https://github.com/prisma/prisma-engines) package, which is a rust-based, compiled binary. The prisma client operates with a graphql based intermediary that gets sent down into the engine for execution, but no real execution details (like the executed query) are sent back.

The best we could do is attach the graphql representation of the query to the span, which would like like the below, but it's not the sql query that you're looking for. Let me know if you've found anything that contradicts what I've said above though!

```gql
query {
  findManyJoke(
    take: 5
    orderBy: [
      {
         createdAt: desc      
      }
    ]
    where: {
      jokesterId: "ab7d0823-4861-47fc-8b35-9b8c5d3fde7f"
    }
  ) {
    id
    name
  }
}
```

**Comment by jcmackie** [comment]
Yeah I was worried that something like this might happen. In my quick perusing of the Prisma repositories and the ongoing tracing work, I saw mentions of tracing in the `prisma/engines` repo: https://github.com/prisma/prisma-engines/pull/3064/files.

I think that what we have so far in this library (`opentelemetry-plugin-prisma-client`), is enough to help guide me in the performance troubleshooting I need to do. It feels like anything more will be impossible because of the Prisma :)

The only last potential option here... is somehow hooking into the events, the way that the `$on` function does: 
https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#on
Apparently if you turn on `query` logging you can capture the actual SQL query information before it's logged to standard out: https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#log-levels

I've tested using the `$on` function to instrument spans... but the duration provided by that function doesn't match what the `request()` function does. Thoughts?

The next thing I need to do is try and make improvements to the OTel Collector X-ray exporter ;)

**Comment by justindsmith** [comment]
Following back up on this. I was able to spend some time looking at the events (thanks for the pointer!), I'm not sure they will give you exactly what you're looking for. While they do report an executed query, because the requests into the engine are executed asynchronously, there is no easy way (that I found) in associating a specific query with a specific request. For example, if I request `users.findUnique` and `notes.findMany` back to back, the event system out log out two queries (one for each request) but there is no indication of which is which. I suppose we could assume that the order is the same and pair them that way, but I'm a bit worried that's not _always_ guaranteed to be the case, especially if the engine has to execute multiple queries to join data for a single request (though I'm not 100% sure that actually ever happens?).

At any rate, it seems that we'd have to do the association in the patched code, which could be fragile.

**Comment by jcmackie** [comment]
Ah, thanks for thoroughly investigating @justindsmith. I completely agree with your assessment. I think the solution currently offered by this package is probably the best we're going to be able to get until OTel is officially supported by the upstream Prisma libraries.
I'm going to close this issue now, thanks for all your help.

**Comment by matthewmueller** [comment]
Hey all, I wanted to share that we're shipping Prisma Client tracing soon that will give you more details about execution. 

Here's an early sneak peak of what's going out in our 4.2.0 release next Tuesday: https://deploy-preview-3408--prisma2-docs.netlify.app/docs/concepts/components/prisma-client/opentelemetry-tracing. It's already available in the 4.1.0 codebase if you feel like giving it a go today.

I'd love to get your feedback on our design and hit me up if you give tracing a shot in your app!
